File: codeigniter4-framework-b3359be/app/Models/ChatModel.php
<?php

namespace App\Models;

use CodeIgniter\Model;

class ChatModel extends Model
{
    protected $table            = 'chats';
    protected $primaryKey       = 'id';
    protected $useAutoIncrement = true;
    protected $returnType       = 'array';
    protected $useSoftDeletes   = false;

    protected $allowedFields = [
        'sender_id',
        'receiver_id',
        'message',
        'reply_to_id',
        'is_read',
        'created_at'
    ];

    // Dates
    protected $useTimestamps = false; // Karena di Laravel model-nya tidak punya updated_at, kita tangani manual
    protected $createdField  = 'created_at';
    protected $updatedField  = 'updated_at';

    // Validation
    protected $validationRules      = [];
    protected $validationMessages   = [];
    protected $skipValidation       = false;
    protected $cleanValidationRules = true;

    /**
     * Mengambil riwayat chat antara dua user (User A dan User B)
     *
     * @param int $user1 ID user pertama
     * @param int $user2 ID user kedua
     * @return array
     */
    public function getChatHistory($user1, $user2)
    {
        return $this->groupStart()
                        ->where('sender_id', $user1)
                        ->where('receiver_id', $user2)
                    ->groupEnd()
                    ->orGroupStart()
                        ->where('sender_id', $user2)
                        ->where('receiver_id', $user1)
                    ->groupEnd()
                    ->orderBy('created_at', 'ASC')
                    ->findAll();
    }
}

File: speakapps_mobile/lib/models/chat_model.dart
class ChatModel {
  final int id;
  final int senderId;
  final int receiverId;
  final String message;
  final int? replyToId;
  final bool isRead;
  final String createdAt;
  final String? friendName;
  final String? friendProfilePic;

  ChatModel({
    required this.id,
    required this.senderId,
    required this.receiverId,
    required this.message,
    this.replyToId,
    required this.isRead,
    required this.createdAt,
    this.friendName,
    this.friendProfilePic,
  });

  factory ChatModel.fromJson(Map<String, dynamic> json) {
    return ChatModel(
      id: int.tryParse(json['id'].toString()) ?? 0,
      senderId: int.tryParse(json['sender_id'].toString()) ?? 0,
      receiverId: int.tryParse(json['receiver_id'].toString()) ?? 0,
      message: json['message'] ?? '',
      replyToId: json['reply_to_id'] != null ? int.tryParse(json['reply_to_id'].toString()) : null,
      isRead: json['is_read'] == 1 || json['is_read'] == '1' || json['is_read'] == true,
      createdAt: json['created_at'] ?? '',
      friendName: json['friend_name'],
      friendProfilePic: json['friend_profile_pic'],
    );
  }
}
